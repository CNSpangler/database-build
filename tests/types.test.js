const {
  isNumber,
  isString,
  isBoolean,
  castToNumber,
  castToString,
  castToBoolean,
  getCaster
} = require('../lib/types.js');

// Validate numbers and associated casters
describe('validator module', () => {
  describe('basic validation', () => {
    it('properly tells if a value is a number', () => {
      expect(isNumber(3)).toBeTruthy();
      expect(isNumber('hi')).toBeFalsy();
      expect(isNumber([])).toBeFalsy();
      expect(isNumber({})).toBeFalsy();
      expect(isNumber(() => {})).toBeFalsy();
      expect(isNumber(true)).toBeFalsy();
    });
  });

  describe('casters', () => {
    it('can cast values to a number', () => {
      expect(castToNumber(3)).toEqual(3);
      expect(castToNumber('3')).toEqual(3);
      expect(castToNumber(true)).toEqual(1);
      expect(castToNumber(false)).toEqual(0);
    });

    it('throws if value is not castable to number', () => {
      expect(() => castToNumber('hi')).toThrowErrorMatchingSnapshot();
      expect(() => castToNumber({})).toThrowErrorMatchingSnapshot();
    });
  });

  it('can get the right number caster', () => {
    expect(getCaster(Number)).toEqual(castToNumber);
    expect(getCaster(Promise)).toBeNull();
  });

  // Validate strings and associated casters
  describe('basic validation', () => {
    it('properly tells if a value is a string', () => {
      expect(isString('wow')).toBeTruthy();
      expect(isString(false)).toBeFalsy();
      expect(isString(6)).toBeFalsy();
      expect(isString([])).toBeFalsy();
      expect(isString({})).toBeFalsy();
      expect(isString(() => {})).toBeFalsy();
    });
  });

  describe('casters', () => {
    it('can cast values to a string', () => {
      expect(castToString(3)).toEqual('3');
      expect(castToString('hi')).toEqual('hi');
    });

    it('throws if value is not castable to string', () => {
      expect(() => castToString({}).toThrowErrorMatchingSnapshot()
      );
    });
  });

  it('can get the right stringy caster', () => {
    expect(getCaster(String)).toEqual(castToString);
    expect(getCaster(Promise)).toBeNull();
  });
});

// Validate booleans and associated casters
describe('basic validation', () => {
  it('properly tells if a value is a boolean', () => {
    expect(isBoolean(true)).toBeTruthy();
    expect(isBoolean(false)).toBeTruthy();
    expect(isBoolean('hi')).toBeFalsy();
    expect(isBoolean(8)).toBeFalsy();
    expect(isBoolean([])).toBeFalsy();
    expect(isBoolean({})).toBeFalsy();
    expect(isBoolean(() => {})).toBeFalsy();
  });
});

describe('casters', () => {
  it('can cast values to a boolean', () => {
    expect(castToBoolean('true')).toEqual(true);
    expect(castToBoolean('false')).toEqual(false);
    expect(castToBoolean(1)).toEqual(true);
    expect(castToBoolean(0)).toEqual(false);
  });

  it('throws if value is not castable to boolean', () => {
    expect(() => castToBoolean('hi')).toThrowErrorMatchingSnapshot();
    expect(() => castToBoolean({})).toThrowErrorMatchingSnapshot();
  });
});
