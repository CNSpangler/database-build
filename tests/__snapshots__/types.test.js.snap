// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validator module casters throws if value is not castable to boolean 1`] = `"Maximum call stack size exceeded"`;

exports[`validator module casters throws if value is not castable to boolean 2`] = `"Maximum call stack size exceeded"`;

exports[`validator module casters throws if value is not castable to number 1`] = `"Cannot cast >>hi<< to Number"`;

exports[`validator module casters throws if value is not castable to number 2`] = `"Cannot cast >>[object Object]<< to Number"`;
